### ISSUE TITLE
Support for private packages

### ISSUE THREAD

**Issue by aaronbauman** [pantheon-systems/terminus-build-tools-plugin#275]
I've got a composer base project downstream of https://github.com/pantheon-systems/example-drops-8-composer/ that relies on private packages.
Once I have all my auth in place, everything works great.

However, the first build for my initial spinup fails because of my reliance on private packages. I'm working around this by creating a `COMPOSER_AUTH` variable immediately after `build:project:create` - which is fine, but makes the process more manual than I'd like.

Is there a way I could be doing this to avoid the manual step?

I'm assuming no, but would welcome help in formulating a PR to support this workflow.

**Comment by ataylorme** [comment]
@aaronbauman does exporting `COMPOSER_AUTH` _before_ running `build:project:create` work for you?

Build Tools could potentially do this and store the Composer auth JSON value the same way it stores other auth credentials, such as `BITBUCKET_PASS`.

**Comment by ataylorme** [comment]
In scope:
- Ensue if `COMPOSER_AUTH` is a defined environment variable, it is passed through to the `build:project:create` process
- Allow for reading `COMPOSER_AUTH` from the Terminus configuration yaml file at `~/.terminus/config.yml`
  - If the value is a string, assume a JSON string
  - If the value is not a string, parse the yaml into a JSON object
- Update documentation on how to use Composer auth

Out of scope:
- Prompting the user for a Composer auth token during `build:project:create` interactively

**Comment by aaronbauman** [comment]
Opened https://github.com/pantheon-systems/terminus-build-tools-plugin/pull/277 against this feature request.

I was able to test fetching `COMPOSER_AUTH` from environment, but not sure how the best way to read `~/.terminus/config.yml`

**Comment by danielbachhuber** [comment]
> I was able to test fetching `COMPOSER_AUTH` from environment, but not sure how the best way to read `~/.terminus/config.yml`

What were you stuck on?

**Comment by aaronbauman** [comment]
> > I was able to test fetching `COMPOSER_AUTH` from environment, but not sure how the best way to read `~/.terminus/config.yml`
> 
> What were you stuck on?

I ended up doing:
```
$this->session()->getConfig()->get('COMPOSER_AUTH');
```
Just wasn't sure if that's the recommended way to do it.

**Comment by rvtraveller** [comment]
The `~/.terminus/config.yml` should automatically be read in as part of the config as you are mentioning. Terminus takes care of loading these and merging with the environment variables for use in the application (you can see this in Terminus here: https://github.com/pantheon-systems/terminus/blob/master/bin/terminus#L38). You'll notice that the order is constants, config file, .env file in your current directory, and then environment variables.

If you are working with a configuration setting in your config.yml that is at the top level, this all works fine and good. However, there is a bug in Terminus where environment variables aren't properly expanding to override nested configuration options. In theory, dots and dashes should be equivalent -- `$this->session()->getConfig()->get('TERMINUS_BUILD-TOOLS_COMPOSER-AUTH')` should be equivalent to `$this->session()->getConfig()->get('terminus.build-tools.composer-auth')` (with the latter being what you need to do to refer to things in your `config.yml`.

As such, I've had to implement elsewhere a double check where I try and retrieve from `config.yml` first (using the dot notation) and following it up by doing a `getenv()` call to retrieve the same environment variable.

Does this make sense?

**Comment by ataylorme** [comment]
> However, there is a bug in Terminus where environment variables aren't properly expanding to override nested configuration options.

It sounds like the best approach would be to fix the bug upstream in Terminus, rather than continuing to work around it.

**Comment by ataylorme** [comment]
I opened an issue on Terminus and [left some notes there](https://github.com/pantheon-systems/terminus/issues/2017#issuecomment-537126515)

tl;dr is we should not be using hyphens. `TERMINUS_BUILD_TOOLS_COMPOSER_AUTH` maps to `build_tools_composer_auth` correctly.

You can verify with `export TERMINUS_BUILD_TOOLS_COMPOSER_AUTH="SEEME?" && terminus self:config:dump | grep -i -C 2 "composer"`
